server/controllers/user_controller.js

```diff
@@ -33,3 +30,58 @@ export const updateUserExploreNumber = (userId, value) => {
     return user.save();
   });
 };
+
+// Notifications are always prepended to the array, so the newest ones appear first.
+export const addUserNotification = (userId, type, sender, href) => {
+  if (!NOTIFICATION_TYPES.includes(type)) {
+    return Promise.reject(new Error('Unrecognized notification type'));
+  }
+  const notification = { messageType: type };
+  if (sender) notification.sender = sender;
+  if (href) notification.href = href;
+  return User.findByIdAndUpdate(
+    userId,
+    { $push: { notifications: { $each: [notification], $position: 0 } } },
+    { new: true, select: 'notifications' }
+  );
+};
+
+export const setNotificationsRead = (userId, notificationIds) => {
+  return User.findById(userId, 'notifications')
+  .then((user) => {
+    for (const id of notificationIds) {
+      const notification = user.notifications.id(id);
+      notification.isRead = true;
+    }
+
+    return user.save();
+  });
+};
+
+/* limit: # of notifications to return
+* page: # of sets of [limit] notifications to skip from start of array
+* automatically marks retrieved notifications as read unless "noRead" is truthy.
+*/
+export const getUserNotifications = (userId, limit, page, noRead) => {
+  page = page || 0;
+  let skip = 0;
+  let projection;
+  if (typeof limit === 'number') {
+    skip = page * limit;
+    projection = { notifications: { $slice: [skip, limit] } };
+  } else {
+    projection = { notifications: 1 };
+  }
+
+  return User.findById(userId, projection)
+  .then((user) => {
+    // mark unread notifications as read
+    if (!noRead) {
```

Comment:
Currently all annotations are set as read when they are viewed which I think is what we want